### PR TITLE
remove ParameterList alias for parameters

### PR DESCRIPTION
**Pull request by WalterBright**
Slowly improving the encapsulation of parameters.

### CODE REVIEW

File: src/dmd/doc.d [hunk 1]
```diff
@@ -2309,9 +2310,9 @@ private TypeFunction isTypeFunction(Dsymbol s)
 private Parameter isFunctionParameter(Dsymbol s, const(char)* p, size_t len)
 {
     TypeFunction tf = isTypeFunction(s);
-    if (tf && tf.parameters)
+    if (tf && tf.parameterList.parameters)
     {
-        foreach (fparam; *tf.parameters)
+        foreach (fparam; *tf.parameterList.parameters)
```
Comment (thewilsonator) [inline src/dmd/doc.d]: `foreach (fparam; tf.parameterList[])`?

File: src/dmd/dtemplate.d [hunk 2]
```diff
@@ -951,8 +951,8 @@ extern (C++) final class TemplateDeclaration : ScopeDsymbol
                 fd.inferRetType = true;
 
                 // Shouldn't run semantic on default arguments and return type.
-                for (size_t i = 0; i < tf.parameters.dim; i++)
-                    (*tf.parameters)[i].defaultArg = null;
+                for (size_t i = 0; i < tf.parameterList.parameters.dim; i++)
+                    (*tf.parameterList.parameters)[i].defaultArg = null;
```
Comment (thewilsonator) [inline src/dmd/dtemplate.d]: foreach

File: src/dmd/dtemplate.d [hunk 3]
```diff
@@ -2164,8 +2164,8 @@ extern (C++) final class TemplateDeclaration : ScopeDsymbol
         Scope* scx = sc2.push();
 
         // Shouldn't run semantic on default arguments and return type.
-        for (size_t i = 0; i < tf.parameters.dim; i++)
-            (*tf.parameters)[i].defaultArg = null;
+        for (size_t i = 0; i < tf.parameterList.parameters.dim; i++)
+            (*tf.parameterList.parameters)[i].defaultArg = null;
```
Comment (thewilsonator) [inline src/dmd/dtemplate.d]: ditto

File: src/dmd/escape.d [hunk 4]
```diff
@@ -992,7 +992,7 @@ private void inferReturn(FuncDeclaration fd, VarDeclaration v)
     else
     {
         // Perform 'return' inference on parameter
-        if (tf.ty == Tfunction && tf.parameters)
+        if (tf.ty == Tfunction)
```
Comment (thewilsonator) [inline src/dmd/escape.d]: This doesn't look like a refactoring change

File: src/dmd/expressionsem.d [hunk 5]
```diff
@@ -4570,9 +4570,7 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
 
                 //printf("tf = %s, args = %s\n", tf.deco, (*arguments)[0].type.deco);
                 .error(exp.loc, "%s `%s%s` is not callable using argument types `%s`",
-                    p, exp.e1.toChars(), parametersTypeToChars(tf.parameters, tf.varargs), buf.peekString());
-                if (failMessage)
-                    errorSupplemental(exp.loc, failMessage);
+                    p, exp.e1.toChars(), parametersTypeToChars(tf.parameterList.parameters, tf.varargs), buf.peekString());
```
Comment (thewilsonator) [inline src/dmd/expressionsem.d]: ditto

File: src/dmd/mtype.d [hunk 6]
```diff
@@ -636,9 +636,9 @@ extern (C++) abstract class Type : RootObject
                 notcovariant |= !fparam1.isCovariant(t1.isref, fparam2);
             }
         }
-        else if (t1.parameters != t2.parameters)
+        else if (t1.parameterList.parameters != t2.parameterList.parameters)
         {
-            if (t1.parameterList.length || t2.parameterList.length)
+            if (t1.parameterList.length || t1.parameterList.length)
```
Comment (thewilsonator) [inline src/dmd/mtype.d]: Nope.

File: src/dmd/parse.d [hunk 7]
```diff
@@ -4455,9 +4455,9 @@ final class Parser(AST) : Lexer
                             assert(tpl_ !is null && tpl_.members.dim == 1);
                             auto fd = cast(AST.FuncLiteralDeclaration) (*tpl_.members)[0];
                             auto tf = cast(AST.TypeFunction) fd.type;
-                            assert(tf.parameters.dim > 0);
+                            assert(tf.parameterList.parameters.dim > 0);
                             auto as = new AST.Dsymbols();
-                            (*tf.parameters)[0].userAttribDecl = new AST.UserAttributeDeclaration(udas, as);
+                            (*tf.parameterList.parameters)[0].userAttribDecl = new AST.UserAttributeDeclaration(udas, as);
```
Comment (thewilsonator) [inline src/dmd/parse.d]: `tf.parameterList[0]`

File: src/dmd/semantic3.d [hunk 8]
```diff
@@ -463,11 +463,11 @@ private extern(C++) final class Semantic3Visitor : Visitor
...
             // Declare the tuple symbols and put them in the symbol table,
             // but not in parameters[].
-            if (f.parameters)
+            if (f.parameterList.parameters)
             {
-                for (size_t i = 0; i < f.parameters.dim; i++)
+                for (size_t i = 0; i < f.parameterList.parameters.dim; i++)
                 {
-                    Parameter fparam = (*f.parameters)[i];
+                    Parameter fparam = (*f.parameterList.parameters)[i];
```
Comment (thewilsonator) [inline src/dmd/semantic3.d]: `foreach`?

File: src/dmd/typesem.d [hunk 9]
```diff
@@ -1120,14 +1120,14 @@ extern(C++) Type typeSemantic(Type t, Loc loc, Scope* sc)
          * as semantic() will get called again on this.
          */
         TypeFunction tf = mtype.copy().toTypeFunction();
-        if (mtype.parameters)
+        if (mtype.parameterList.parameters)
         {
-            tf.parameters = mtype.parameters.copy();
-            for (size_t i = 0; i < mtype.parameters.dim; i++)
+            tf.parameterList.parameters = mtype.parameterList.parameters.copy();
+            for (size_t i = 0; i < mtype.parameterList.parameters.dim; i++)
```
Comment (thewilsonator) [inline src/dmd/typesem.d]: `foreach`

File: test/fail_compilation/fail99.d [hunk 10]
```diff
@@ -2,10 +2,10 @@
 TEST_OUTPUT:
 ---
 fail_compilation/fail99.d(13): Error: delegate `dg(int)` is not callable using argument types `()`
-fail_compilation/fail99.d(13):        missing argument for parameter #1: `int`
```
Comment (thewilsonator) [inline test/fail_compilation/fail99.d]: ?

**Review by thewilsonator — CHANGES_REQUESTED** [review]
Otherwise looks good

File: src/dmd/escape.d (hunk 4 above)
Comment (WalterBright) [reply src/dmd/escape.d]: It's covered by line 997, where dim will be 0 for !tf.parameters, and the loop will execute 0 times.

Comment (thewilsonator) [reply src/dmd/escape.d]: OK

File: src/dmd/expressionsem.d (hunk 5 above)
Comment (WalterBright) [reply src/dmd/expressionsem.d]: ??

Comment (thewilsonator) [reply src/dmd/expressionsem.d]: "This doesn't look like a refactoring change"

File: test/fail_compilation/fail99.d (hunk 10 above)
Comment (thewilsonator) [reply test/fail_compilation/fail99.d]: I think that was part of this.

File: src/dmd/parse.d (hunk 7 above)
Comment (WalterBright) [reply src/dmd/parse.d]: that has different semantics

File: src/dmd/semantic3.d (hunk 8 above)
Comment (WalterBright) [reply src/dmd/semantic3.d]: later

File: src/dmd/typesem.d (hunk 9 above)
Comment (WalterBright) [reply src/dmd/typesem.d]: later

File: src/dmd/expressionsem.d (hunk 5 above)
Comment (WalterBright) [reply src/dmd/expressionsem.d]: Oh, I see. I accidentally deleted a couple lines.

**Comment by WalterBright** [comment]
I've had enough trouble with this seemingly simple PR that going any further will be in a later PR. It amply illustrates a failure of encapsulation I'm trying to fix.

**Comment by thewilsonator** [comment]
OK.

**Review by thewilsonator — APPROVED** [review]

**Comment by WalterBright** [comment]
It is unrelated. Please turn automerge back on!

**Pull request merged**